projects/continuous_learning/dend_kwinners.py

```diff
@@ -0,0 +1,195 @@
+                self.n, self.dpc, self.boost_strength,
+                self.boost_strength_factor, self.k_inference_factor,
+                self.duty_cycle_period
+            )
+        )
+
+    @abc.abstractmethod
+    def update_duty_cycle(self, x):
+        r"""Updates our duty cycle estimates with the new value. Duty cycles are
+        updated according to the following formula:
+
+        .. math::
+            dutyCycle = \frac{dutyCycle \times \left( period - batchSize \right)
+                                + newValue}{period}
+
+        :param x:
+          Current activity of each unit
+        """
+        raise NotImplementedError
+
+    def update_boost_strength(self):
+        """Update boost strength by multiplying by the boost strength factor.
+        This is typically done during training at the beginning of each epoch.
+        """
+        self.boost_strength *= self.boost_strength_factor
+
+    def entropy(self):
+        """Returns the current total entropy of this layer."""
+        _, entropy = binary_entropy(self.duty_cycle)
+        return entropy
+
+    def max_entropy(self):
+        """Returns the maximum total entropy we can expect from this layer."""
+        return max_entropy(self.n, int(self.n * self.percent_on))
+
+
+class DKWinners(DKWinnersBase):
+    def __init__(
+        self,
+        n,
+        out_dim,
+        dpc,
+        k_inference_factor=1.5,
+        boost_strength=1.0,
+        boost_strength_factor=0.9,
+        duty_cycle_period=1000,
+    ):
+        super(DKWinners, self).__init__(
+            out_dim=out_dim,
+            dpc=dpc,
+            k_inference_factor=k_inference_factor,
+            boost_strength=boost_strength,
+            boost_strength_factor=boost_strength_factor,
+            duty_cycle_period=duty_cycle_period,
+        )
+        self.n = n
+        self.k = out_dim  # int(round(n * percent_on))
+        self.k_inference = int(self.k * self.k_inference_factor)
+        self.register_buffer("duty_cycle", torch.zeros(self.n))
+
+        if type(self.dpc) == tuple:
+            self.dpc = self.dpc[0]
+
+    def forward(self, x):
+
+        if self.training:
+            x = GenericKWinners.apply(x, self.duty_cycle, self.k,
+                                      self.dpc, self.boost_strength)
+            self.update_duty_cycle(x)
+        else:
+            x = GenericKWinners.apply(x, self.duty_cycle, self.k_inference,
+                                      self.dpc, self.boost_strength)
+
+        return x
+
+    def update_duty_cycle(self, x):
+        batch_size = x.shape[0]
+        self.learning_iterations += batch_size
+        period = min(self.duty_cycle_period, self.learning_iterations)
+        self.duty_cycle.mul_(period - batch_size)
+        self.duty_cycle.add_(x.gt(0).sum(dim=0, dtype=torch.float))
+        self.duty_cycle.div_(period)
+
+
+class GenericKWinners(torch.autograd.Function):
```

Comment:
As discussed, I think our KWinnersLocal might be able to handle this already. If not, you should definitely have  some good unit tests of this function. It's hard for me to tell whether the code is correct.